File: leetcode/815_Bus_Routes.py
from typing import List
from unittest import TestCase
from itertools import combinations
from collections import deque

class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        if source == target:
            return 0
        bus_n = len(routes)
        stop_bus_dict = {}
        adj_list = [[] for i in range(bus_n)]

        for bus in range(bus_n):
            for stop in routes[bus]:
                if not stop in stop_bus_dict:
                    stop_bus_dict[stop] = []
                stop_bus_dict[stop].append(bus)

        for buses in stop_bus_dict.values():
            for v, u in combinations(buses, 2):
                adj_list[v].append(u)
                adj_list[u].append(v)

        if target in stop_bus_dict:
            targets = set(stop_bus_dict[target])
            queue = deque((bus, 1) for bus in stop_bus_dict[source])
            expanded = [False] * bus_n
            while queue:
                v, cost = queue.popleft()
                if not expanded[v]:
                    expanded[v] = True
                    if v in targets:
                        return cost
                    for u in adj_list[v]:
                        if not expanded[u]:
                            queue.append((u, cost + 1))
        return -1

        

class NumBusesToDestinationTest(TestCase):
    def test_1(self) -> None:
        self.assertEqual(2, Solution().numBusesToDestination([[1,2,7],[3,6,7]], 1, 6))

    def test_2(self) -> None:
        self.assertEqual(-1, Solution().numBusesToDestination([[7,12],[4,5,15],[6],[15,19],[9,12,13]], 15, 12))        

File: leetcode/815_Bus_Routes.rs
use std::collections::{HashMap, VecDeque, HashSet};


struct Solution {}

impl Solution {
    pub fn num_buses_to_destination(routes: Vec<Vec<i32>>, source: i32, target: i32) -> i32 {
        if source == target {
            return 0;
        }

        let (source, target) = (source as usize, target as usize);
        let bus_n = routes.len();        
        let mut stop_bus_dic = HashMap::new();
        let mut adj_list = vec![Vec::new(); bus_n];

        for bus in 0..bus_n {
            for stop in routes[bus].iter().map(|x| *x as usize) {
                stop_bus_dic.entry(stop).or_insert(vec![]).push(bus);
            }
        }
        for buses in stop_bus_dic.values() {
            for i in 0..buses.len() {
                for j in i+1..buses.len() {
                    adj_list[buses[i]].push(buses[j]);
                    adj_list[buses[j]].push(buses[i]);
                }
            }
        }
        if let (Some(target_buses), Some(source_buses)) = (stop_bus_dic.get(&target), stop_bus_dic.get(&source)) {
            let targets = target_buses.iter().copied().collect::<HashSet<_>>();
            let mut queue = source_buses.iter().map(|bus| (*bus, 1)).collect::<VecDeque<_>>();
            let mut expended = vec![false; bus_n];
            while let Some((v, cost)) = queue.pop_front() {
                if targets.contains(&v) {
                    return cost;
                }
                if ! expended[v] {
                    expended[v] = true;
                    for u in &adj_list[v] {
                        if ! expended[*u] {
                            queue.push_back((*u, cost + 1));
                        }
                    }
                }
            }
        }
        -1
    }
}

#[test]
fn test_1() {
    assert_eq!(2, Solution::num_buses_to_destination(vec![vec![1,2,7], vec![3,6,7]], 1, 6));
}

#[test]
fn test_2() {
    assert_eq!(-1, Solution::num_buses_to_destination(vec![vec![7,12], vec![4,5,15], vec![6], vec![15,19], vec![9,12,13]], 15, 12));
}
